cupy/_core/_routines_manipulation.pyx

```diff
@@ -594,7 +594,11 @@ cpdef _ndarray_base concatenate_method(
     if dtype is not None:
         dtype = get_dtype(dtype)
 
-    arrays = list(tup)
+    arrays = []
+    for array in tup:
+        if hasattr(array, '__cupy_get_ndarray__'):
```

Comment:
Should we consider `__array__` as well?

Reply:
Any ideas on this one?

Reply:
This seems inconsistent in numpy right now:
```python
>>> import cupy as cp
>>> import numpy as np
>>> import torch
>>> import numpy.array_api

>>> t = torch.tensor([1, 2], device='cpu')  # has __array__, not buffer protocol
>>> np.array_api.asarray(t)
Array([1, 2], dtype=int64)
>>> np.array_api.asarray(x)
Array([0, 1], dtype=int64)
>>> np.array_api.asarray([x, x])  # via buffer protocol says the array_api.asarray docstring
Array([[0, 1],
       [0, 1]], dtype=int64)

>>> # The issue:
>>> np.array_api.asarray([t, t])
/home/rgommers/anaconda3/envs/cupy-torch/lib/python3.9/site-packages/numpy/array_api/_creation_functions.py:71: FutureWarning: The input object of type 'Tensor' is an array-like implementing one of the corresponding protocols (`__array__`, `__array_interface__` or `__array_struct__`); but not a sequence (or 0-D). In the future, this object will be coerced as if it was first converted using `np.array(obj)`. To retain the old behaviour, you have to either modify the type 'Tensor', or assign to an empty array created with `np.empty(correct_shape, dtype=object)`.
  res = np.asarray(obj, dtype=dtype)
/home/rgommers/anaconda3/envs/cupy-torch/lib/python3.9/site-packages/numpy/array_api/_creation_functions.py:71: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  res = np.asarray(obj, dtype=dtype)
Traceback (most recent call last):
  Input In [31] in <cell line: 1>
    np.array_api.asarray([t, t])
  File ~/anaconda3/envs/cupy-torch/lib/python3.9/site-packages/numpy/array_api/_creation_functions.py:72 in asarray
    return Array._new(res)
  File ~/anaconda3/envs/cupy-torch/lib/python3.9/site-packages/numpy/array_api/_array_object.py:80 in _new
    raise TypeError(
TypeError: The array_api namespace does not support the dtype 'object'
```

So:
-  `__array__` is used for converting a single array
- a sequence where elements support the buffer protocol are okay too
- but, a sequence where elements support `__array__` is not.

Unclear if that is on purpose. @asmeurer can you comment?

Reply:
Does `np.asarray` work here? np.asarray just calls out to that.

Reply:
Kinda but not really:
```
>>> np.asarray([t, t])
<ipython-input-33-035f9ee92d0b>:1: FutureWarning: The input object of type 'Tensor' is an array-like implementing one of the corresponding protocols (`__array__`, `__array_interface__` or `__array_struct__`); but not a sequence (or 0-D). In the future, this object will be coerced as if it was first converted using `np.array(obj)`. To retain the old behaviour, you have to either modify the type 'Tensor', or assign to an empty array created with `np.empty(correct_shape, dtype=object)`.
  np.asarray([t, t])
<ipython-input-33-035f9ee92d0b>:1: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  np.asarray([t, t])
array([tensor([1, 2]), tensor([1, 2])], dtype=object)
```

Reply:
Well that's all it's doing is calling np.asarray, and when it sees that the result is an object array, it hits the guard against that and gives the corresponding error. So I'd say the bug here is that np.asarray doesn't recursively call asarray on sequences. Surely this has been discussed before.

Reply:
I'm resolving this one in the meantime due a lack of clarity about this path on NumPy. Given the scope of this PR, only CuPy array-api objects will get composed